Using `JPA` helper class stand alone. without injecting `JPAApi`

### Play Version (2.6.0)

### API Java)

### Operating System (Ubuntu 17.04)

### JDK (Oracle 1.8.0_131)

### Expected Behavior

Using `JPA` helper class stand alone. without injecting `JPAApi`


### Actual Behavior
**It seems that creating entity manager using DI has some problems in play 2.6**
I have a very very simple database access :
```
@Transactional
public class HomeController extends Controller {

    public Result index() {
        JPA.em().find(Person.class, 1L);
        return ok(views.html.index.render());
    }

}
```

this simple code generates following stack trace and does not works:
```

! @74fhap05p - Internal server error, for (GET) [/] ->
 
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[ProvisionException: Unable to provision, see the following errors:

1) Error injecting constructor, javax.persistence.PersistenceException: Unable to build entity manager factory
  at play.db.jpa.DefaultJPAApi$JPAApiProvider.<init>(DefaultJPAApi.java:40)
  at play.db.jpa.DefaultJPAApi$JPAApiProvider.class(DefaultJPAApi.java:35)
  while locating play.db.jpa.DefaultJPAApi$JPAApiProvider
  while locating play.db.jpa.JPAApi
    for the 1st parameter of play.db.jpa.TransactionalAction.<init>(TransactionalAction.java:20)
  while locating play.db.jpa.TransactionalAction

1 error]]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:255)
        at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:182)
        at play.core.server.AkkaHttpServer$$anonfun$$nestedInanonfun$executeHandler$1$1.applyOrElse(AkkaHttpServer.scala:230)
        at play.core.server.AkkaHttpServer$$anonfun$$nestedInanonfun$executeHandler$1$1.applyOrElse(AkkaHttpServer.scala:229)
        at scala.concurrent.Future.$anonfun$recoverWith$1(Future.scala:412)
        at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
        at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:70)
        at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:68)
        at scala.concurrent.impl.Promise$KeptPromise$Kept.onComplete(Promise.scala:368)
Caused by: com.google.inject.ProvisionException: Unable to provision, see the following errors:

1) Error injecting constructor, javax.persistence.PersistenceException: Unable to build entity manager factory
  at play.db.jpa.DefaultJPAApi$JPAApiProvider.<init>(DefaultJPAApi.java:40)
  at play.db.jpa.DefaultJPAApi$JPAApiProvider.class(DefaultJPAApi.java:35)
  while locating play.db.jpa.DefaultJPAApi$JPAApiProvider
  while locating play.db.jpa.JPAApi
    for the 1st parameter of play.db.jpa.TransactionalAction.<init>(TransactionalAction.java:20)
  while locating play.db.jpa.TransactionalAction

1 error
        at com.google.inject.internal.InjectorImpl$2.get(InjectorImpl.java:1028)
        at com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1054)
        at play.api.inject.guice.GuiceInjector.instanceOf(GuiceInjectorBuilder.scala:405)
        at play.core.j.DefaultJavaHandlerComponents.getAction(JavaAction.scala:173)
        at play.core.j.JavaAction.$anonfun$apply$1(JavaAction.scala:96)
        at scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:56)
        at scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:64)
        at scala.collection.mutable.ArraySeq.foldLeft(ArraySeq.scala:46)
        at play.core.j.JavaAction.apply(JavaAction.scala:94)
        at play.api.mvc.Action.$anonfun$apply$2(Action.scala:96)
Caused by: javax.persistence.PersistenceException: Unable to build entity manager factory
        at org.hibernate.jpa.HibernatePersistenceProvider.createEntityManagerFactory(HibernatePersistenceProvider.java:66)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:39)
        at play.db.jpa.DefaultJPAApi.lambda$start$1(DefaultJPAApi.java:61)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at play.db.jpa.DefaultJPAApi.start(DefaultJPAApi.java:60)
        at play.db.jpa.DefaultJPAApi$JPAApiProvider.<init>(DefaultJPAApi.java:47)
        at play.db.jpa.DefaultJPAApi$JPAApiProvider$$FastClassByGuice$$dcd4cdbd.newInstance(<generated>)
        at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:89)
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:111)
Caused by: org.hibernate.boot.registry.classloading.spi.ClassLoadingException: Unable to load class [models.Person]
        at org.hibernate.boot.registry.classloading.internal.ClassLoaderServiceImpl.classForName(ClassLoaderServiceImpl.java:229)
        at org.hibernate.boot.model.source.internal.annotations.AnnotationMetadataSourceProcessorImpl.<init>(AnnotationMetadataSourceProcessorImpl.java:103)
        at org.hibernate.boot.model.process.spi.MetadataBuildingProcess$1.<init>(MetadataBuildingProcess.java:147)
        at org.hibernate.boot.model.process.spi.MetadataBuildingProcess.complete(MetadataBuildingProcess.java:141)
        at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.metadata(EntityManagerFactoryBuilderImpl.java:848)
        at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:876)
        at org.hibernate.jpa.HibernatePersistenceProvider.createEntityManagerFactory(HibernatePersistenceProvider.java:58)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:39)
        at play.db.jpa.DefaultJPAApi.lambda$start$1(DefaultJPAApi.java:61)
Caused by: java.lang.ClassNotFoundException: Could not load requested class : models.Person
        at org.hibernate.boot.registry.classloading.internal.ClassLoaderServiceImpl$AggregatedClassLoader.findClass(ClassLoaderServiceImpl.java:217)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at org.hibernate.boot.registry.classloading.internal.ClassLoaderServiceImpl.classForName(ClassLoaderServiceImpl.java:226)
        at org.hibernate.boot.model.source.internal.annotations.AnnotationMetadataSourceProcessorImpl.<init>(AnnotationMetadataSourceProcessorImpl.java:103)
        at org.hibernate.boot.model.process.spi.MetadataBuildingProcess$1.<init>(MetadataBuildingProcess.java:147)
        at org.hibernate.boot.model.process.spi.MetadataBuildingProcess.complete(MetadataBuildingProcess.java:141)
        at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.metadata(EntityManagerFactoryBuilderImpl.java:848)


```

while, surprisingly adding unused `JPAApi` inject resolves this problem:
```
@Transactional
public class HomeController extends Controller {
    @Inject
    public HomeController(JPAApi jpaApi) {
    }
    public Result index() {
        JPA.em().find(Person.class, 1L);
        return ok(views.html.index.render());
    }
}
```

I meet this issue when I am try to update my play version from 2.5 to 2.6

Hi @mohsenkashi, some questions:

1. Is it possible to run a `sbt clean` so that we can confirm this is not being caused by previous compilations with another Play version? The root cause here is this:

```Caused by: java.lang.ClassNotFoundException: Could not load requested class : models.Person
        at org.hibernate.boot.registry.classloading.internal.ClassLoaderServiceImpl$AggregatedClassLoader.findClass(ClassLoaderServiceImpl.java:217)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at org.hibernate.boot.registry.classloading.internal.ClassLoaderServiceImpl.classForName(ClassLoaderServiceImpl.java:226)
        at org.hibernate.boot.model.source.internal.annotations.AnnotationMetadataSourceProcessorImpl.<init>(AnnotationMetadataSourceProcessorImpl.java:103)
        at org.hibernate.boot.model.process.spi.MetadataBuildingProcess$1.<init>(MetadataBuildingProcess.java:147)
        at org.hibernate.boot.model.process.spi.MetadataBuildingProcess.complete(MetadataBuildingProcess.java:141)
        at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.metadata(EntityManagerFactoryBuilderImpl.java:848)
```
So, I'm not sure this is related to the use of `JPA` or something else.

2.  If `sbt clean` does not solve the problem, could you please create a small project that shows this problem?

Thanks.

--- a/framework/src/play-guice/src/main/scala/play/api/inject/guice/GuiceInjectorBuilder.scala
+++ b/framework/src/play-guice/src/main/scala/play/api/inject/guice/GuiceInjectorBuilder.scala
@@ -5,11 +5,13 @@ package play.api.inject
 package guice
 
 import com.google.inject.util.{ Modules => GuiceModules, Providers => GuiceProviders }
-import com.google.inject.{ Module => GuiceModule, Binder, Stage, CreationException, Guice }
+import com.google.inject.{ Binder, CreationException, Guice, Stage, Module => GuiceModule }
 import java.io.File
-import javax.inject.Inject
+import javax.inject.{ Inject, Provider }
+
 import play.api.inject.{ Binding => PlayBinding, Injector => PlayInjector, Module => PlayModule }
 import play.api.{ Configuration, Environment, Mode, PlayException }
+
 import scala.collection.JavaConverters._
 import scala.reflect.ClassTag
 
@@ -157,7 +159,9 @@ abstract class GuiceBuilder[Self] protected (
   def createModule(): GuiceModule = {
     import scala.collection.JavaConverters._
     val injectorModule = GuiceableModule.guice(Seq(
-      bind[PlayInjector].to[GuiceInjector],
+      bind[GuiceInjector].toSelf,
+      bind[GuiceClassLoader].to(new GuiceClassLoader(environment.classLoader)),
+      bind[PlayInjector].toProvider[GuiceInjectorWithClassLoaderProvider],
       // Java API injector is bound here so that it's available in both
       // the default application loader and the Java Guice builders
       bind[play.inject.Injector].to[play.inject.DelegateInjector]
@@ -408,4 +412,24 @@ class GuiceInjector @Inject() (injector: com.google.inject.Injector) extends Pla
    * Get an instance bound to the given binding key.
    */
   def instanceOf[T](key: BindingKey[T]) = injector.getInstance(GuiceKey(key))
+}
+
+/**
+ * An object that holds a `ClassLoader` for Guice to use. We use this
+ * simple value object so it can be looked up by its type when we're
+ * assembling the Guice injector.
+ *
+ * @param classLoader The wrapped `ClassLoader`.
+ */
+class GuiceClassLoader(val classLoader: ClassLoader)
+
+/**
+ * A provider for a Guice injector that wraps the injector to ensure
+ * it uses the correct `ClassLoader`.
+ *
+ * @param injector The injector to wrap.
+ * @param guiceClassLoader The `ClassLoader` the injector should use.
+ */
+class GuiceInjectorWithClassLoaderProvider @Inject() (injector: GuiceInjector, guiceClassLoader: GuiceClassLoader) extends Provider[Injector] {
+  override def get(): PlayInjector = new ContextClassLoaderInjector(injector, guiceClassLoader.classLoader)
 }

--- a/framework/src/play-guice/src/test/scala/play/api/inject/guice/GuiceInjectorBuilderSpec.scala
+++ b/framework/src/play-guice/src/test/scala/play/api/inject/guice/GuiceInjectorBuilderSpec.scala
@@ -10,7 +10,7 @@ import java.net.URLClassLoader
 import com.google.inject.AbstractModule
 import org.specs2.mutable.Specification
 import play.api.inject._
-import play.api.{ Configuration, Environment, Mode }
+import play.api.{ Configuration, Environment, Mode, inject }
 
 class GuiceInjectorBuilderSpec extends Specification {
 
@@ -37,6 +37,16 @@ class GuiceInjectorBuilderSpec extends Specification {
       env.rootPath must_== new File("test")
       env.mode must_== Mode.Dev
       env.classLoader must be(classLoader)
+    }
+
+    "use the right ClassLoader when injecting" in {
+      val classLoader = new URLClassLoader(Array.empty)
+      val classLoaderAware = new GuiceInjectorBuilder()
+        .in(classLoader)
+        .bindings(bind[GuiceInjectorBuilderSpec.ClassLoaderAware].toSelf)
+        .injector().instanceOf[GuiceInjectorBuilderSpec.ClassLoaderAware]
+
+      classLoaderAware.constructionClassLoader must_== classLoader
     }
 
     "set configuration" in {
@@ -143,6 +153,11 @@ object GuiceInjectorBuilderSpec {
 
   class SetConfigurationModule(conf: Configuration) extends AbstractModule {
     def configure() = bind(classOf[Configuration]) toInstance conf
+  }
+
+  class ClassLoaderAware {
+    // This is the value of the Thread's context ClassLoader at the time the object is constructed
+    val constructionClassLoader: ClassLoader = Thread.currentThread.getContextClassLoader
   }
 
   trait A

--- a/framework/src/play/src/main/scala/play/api/Application.scala
+++ b/framework/src/play/src/main/scala/play/api/Application.scala
@@ -267,12 +267,13 @@ trait BuiltInComponents extends I18nComponents {
    * existing (deprecated) legacy APIs to function. It is not set up to support injecting arbitrary Play components.
    */
   lazy val injector: Injector = {
-    new SimpleInjector(NewInstanceInjector) +
+    val simple = new SimpleInjector(NewInstanceInjector) +
       cookieSigner + // play.api.libs.Crypto (for cookies)
       httpConfiguration + // play.api.mvc.BodyParsers trait
       tempFileCreator + // play.api.libs.TemporaryFileCreator object
       messagesApi + // play.api.i18n.Messages object
       langs // play.api.i18n.Langs object
+    new ContextClassLoaderInjector(simple, environment.classLoader)
   }
 
   lazy val playBodyParsers: PlayBodyParsers =

--- a/framework/src/play/src/main/scala/play/api/inject/Injector.scala
+++ b/framework/src/play/src/main/scala/play/api/inject/Injector.scala
@@ -109,3 +109,21 @@ class SimpleInjector(fallback: Injector, components: Map[Class[_], Any] = Map.em
     new SimpleInjector(fallback, components + (clazz -> component))
 
 }
+
+/**
+ * Wraps an existing injector, ensuring all calls have the correct context `ClassLoader` set.
+ */
+private[play] class ContextClassLoaderInjector(delegate: Injector, classLoader: ClassLoader) extends Injector {
+  override def instanceOf[T: ClassManifest]: T = withContext { delegate.instanceOf[T] }
+  override def instanceOf[T](clazz: Class[T]): T = withContext { delegate.instanceOf(clazz) }
+  override def instanceOf[T](key: BindingKey[T]): T = withContext { delegate.instanceOf(key) }
+
+  @inline
+  private def withContext[T](body: => T): T = {
+    val thread = Thread.currentThread()
+    val oldClassLoader = thread.getContextClassLoader
+    thread.setContextClassLoader(classLoader)
+    try body finally thread.setContextClassLoader(oldClassLoader)
+  }
+
+}

--- a/framework/src/play/src/test/scala/play/api/BuiltInComponentsSpec.scala
+++ b/framework/src/play/src/test/scala/play/api/BuiltInComponentsSpec.scala
@@ -1,0 +1,40 @@
+/*
+ * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
+ */
+package play.api
+
+import java.io.File
+import java.net.URLClassLoader
+
+import org.specs2.mutable.Specification
+import play.api.inject.DefaultApplicationLifecycle
+import play.api.mvc.EssentialFilter
+import play.api.routing.Router
+import play.core.{ SourceMapper, WebCommands }
+
+class BuiltInComponentsSpec extends Specification {
+  "BuiltinComponents" should {
+    "use the Environment ClassLoader for runtime injection" in {
+      val classLoader = new URLClassLoader(Array())
+      val components = new BuiltInComponents {
+        override val environment: Environment = Environment(new File("."), classLoader, Mode.Test)
+        override def configuration: Configuration = Configuration.load(environment)
+        override def applicationLifecycle: DefaultApplicationLifecycle = new DefaultApplicationLifecycle
+        override def router: Router = ???
+        override def webCommands: WebCommands = ???
+        override def sourceMapper: Option[SourceMapper] = ???
+        override def httpFilters: Seq[EssentialFilter] = ???
+      }
+      components.environment.classLoader must_== classLoader
+      val constructedObject = components.injector.instanceOf[BuiltInComponentsSpec.ClassLoaderAware]
+      constructedObject.constructionClassLoader must_== classLoader
+    }
+  }
+}
+
+object BuiltInComponentsSpec {
+  class ClassLoaderAware {
+    // This is the value of the Thread's context ClassLoader at the time the object is constructed
+    val constructionClassLoader: ClassLoader = Thread.currentThread.getContextClassLoader
+  }
+}

